geopandas/geoseries.py

```diff
@@ -777,6 +777,76 @@ def to_crs(self, crs=None, epsg=None):
             GeometryArray(new_data), crs=crs, index=self.index, name=self.name
         )
 
+    def estimate_utm_crs(self, datum_name="WGS 84"):
+        """Returns the estimated UTM CRS based on the bounds of the dataset.
+
+        .. versionadded:: 0.9
+
+        .. note:: Requires pyproj 3+
+
+        Parameters
+        ----------
+        datum_name : str, optional
+            The name of the datum to use in the query. Default is WGS 84.
+
+        Returns
+        -------
+        pyproj.CRS
+
+        Examples
+        --------
+        >>> world = geopandas.read_file(
+        ...     geopandas.datasets.get_path("naturalearth_lowres")
+        ... )
+        >>> germany = world.loc[world.name == "Germany"]
+        >>> germany.geometry.estimate_utm_crs()  # doctest: +SKIP
+        <Projected CRS: EPSG:32632>
+        Name: WGS 84 / UTM zone 32N
+        Axis Info [cartesian]:
+        - E[east]: Easting (metre)
+        - N[north]: Northing (metre)
+        Area of Use:
+        - name: World - N hemisphere - 6°E to 12°E - by country
+        - bounds: (6.0, 0.0, 12.0, 84.0)
+        Coordinate Operation:
+        - name: UTM zone 32N
+        - method: Transverse Mercator
+        Datum: World Geodetic System 1984
+        - Ellipsoid: WGS 84
+        - Prime Meridian: Greenwich
+        """
+        try:
+            from pyproj.aoi import AreaOfInterest
+            from pyproj.database import query_utm_crs_info
+        except ImportError:
+            raise RuntimeError("pyproj 3+ required for estimate_utm_crs.")
+
+        if not self.crs:
+            raise RuntimeError("crs must be set to estimate UTM CRS.")
+
+        # ensure using geographic coordinates
+        if not self.crs.is_geographic:
+            geoseries = self.to_crs("EPSG:4326")
```

Comment:
Just a thought. Would it make sense to reproject only total_bounds in this case and use them directly below? So we avoid potentially costly reprojection of all geometries to 4326. Truth is that I am not entirely certain if it could have some adverse consequences or not.

Reply:
That's a good point to bring up. Reprojecting just the total bounds would be more efficient, but would get a different set of corners in some cases. This would require some investigation to see what the potential issues would be and how much they matter. rasterio has a [transform_bounds](https://rasterio.readthedocs.io/en/latest/api/rasterio.warp.html#rasterio.warp.transform_bounds) method that would be nice to be able to use. But, definitely not a dependency we would want to add.

Reply:
See: https://github.com/corteva/rioxarray/pull/182

Reply:
They densify bounds with additional points before reprojection, that is smart. Probably a bit too complex for us if its going to be used only here, so let's keep it as it is. I guess that people will be using this mostly with 4326 anyway, so the reprojection will not even happen.

Reply:
The reason we densify bounds in rasterio is to help account for projection distortion along the edges between the original total bounds, which would otherwise be ignored by only reprojecting the vertices of the outer bounds.

The actual implementation is [pretty simple](https://github.com/mapbox/rasterio/blob/11ae663ff491092d44a8521e6e7ba33fb920fa22/rasterio/warp.py#L114) and could probably be done even more efficiently with a few lines using numpy (probably [linspace](https://numpy.org/doc/stable/reference/generated/numpy.linspace.html)).

However, for purposes here, you don't need to get the complete outer bounds, you need just enough to come up with a centroid that can be used to find an appropriate UTM zone.  If the distortions are symmetric (e.g., if a function of latitude), then perhaps they are negligible here, since they'd be equal for parallel edges of the bounds.

One possible simplification would be to project the individual bounds of the geometries, then take the outer bounds of this.  It is still an approximation, but possibly faster than projecting all vertices of all geometries first.

Reply:
> I guess that people will be using this mostly with 4326 anyway, so the reprojection will not even happen.

That's likely true. And if it isn't the case, users can do things to optimize the operation by simplifying the dataframe geometry before calling the `estimate_utm_crs` method if they need better performance. If a `transform_bounds`-like method isn't used here, maybe an example of doing so would be useful for the user to consider.

Another bit of code I found here that could also be useful to geopandas users in general to segmentize polygons: https://gist.github.com/underchemist/dab01991bbde83c375f3a83e562fa2fd

I have done something similar in the past to preserve the shape when reprojecting. Mostly for a use case similar to this one.
```python
interp_coords += [
    list(interp_line.interpolate(pt, normalized=True).coords)[0]
    for pt in numpy.linspace(0, 1, densify_pts + 1, endpoint=False)
]
```

If the segmentize method were added to geopandas, then we could:
1. Get the `convex_hull` representing the dataframe (`gdf.unary_union.convex_hull`)
2. Segmentize the convex hull
3. Reproject and get the centroid to estimate the UTM CRS.

However, in my experience I have notice that `unary_union.convex_hull` often times fails due to GEOS not liking the underlying geometry and is not as stable as I would like.

So, we could just use the `total_bounds`, convert it to a box, segmentize, and reproject.

Reply:
Also, I was thinking it is probably okay to merge in its current state. It is stable, just not as performant as we would like for all use cases. Then, we can work on optimizing it in future iterations.

Reply:
Wonder if the segmentize could be added to fiona (unless it is already there and I missed it):
https://gdal.org/python/osgeo.ogr.Geometry-class.html#Segmentize

Looks like postgis has it: https://postgis.net/docs/ST_Segmentize.html, so sounds like a reasonable add to geopandas.

Reply:
That should probably go to pygeos/shapely using GEOS Densifier https://geos.osgeo.org/doxygen/classgeos_1_1geom_1_1util_1_1Densifier.html.

Reply:
Nice find! I agree. I was looking with the wrong word.

Reply:
https://github.com/pygeos/pygeos/issues/258

Reply:
https://github.com/Toblerity/Shapely/issues/1037

Reply:
And the race is on: https://github.com/Toblerity/Fiona/issues/982

Reply:
Probably better in pygeos/shapely, but not sure how fast the GEOS release process is.

Reply:
For purposes of choosing the best UTM zone, is it necessary to preserve the overall shape of the geometries?  It seems like mostly you want a reasonable approximation of the outer bounds of the geometry once projected to geographic coordinates.  Thus densifying the box represented by `total_bounds` is a reasonable, fast approximation.

Also - to some degree extra densification, etc seems unnecessary, if the `total_bounds` fall within a single UTM zone, right?  So another simplification (perhaps this is done in pyproj?) would be to see if the bounds overlap multiple UTM zones; if so, perhaps a bit more work densifying & reprojecting the box may be helpful, and if not, then proceed directly to returning that singular UTM zone.
(these optimizations could be done later)

It might be worthwhile to note in the doctstring that this will reproject all geometries to geographic coordinates if needed; that way the user knows they can sidestep that by using a simplified derived data frame (e.g., bounding boxes of all geometries, densified outer box, etc) to arrive at a faster result.

Reply:
> to some degree extra densification, etc seems unnecessary

The "estimate" part of the function says that we aren't claiming it is the best possible UTM for this dataframe, just one that is reasonable. Way too many ways to complicate this I think (I have thought of a couple since your post that I discarded). With that in mind, we could just transform the bounding coordinates directly with pyproj and call it done :smile: .

Reply:
>  we could just transform the bounding coordinates directly with pyproj and call it done

That sounds reasonable 👍